File: FrondEnd/src/app/core/servicios/dicta.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { AuthServiceService } from './auth.service.service';
import { Dicta, PageResponse } from '../models/auth-response.model'; // Ajusta la ruta

@Injectable({
  providedIn: 'root'
})
export class DictaService {
  private apiURL = `${environment.apiURL}/api/dicta`;
  private http = inject(HttpClient);
  private authService = inject(AuthServiceService);

  private getAuthHeaders(): HttpHeaders {
    const token = this.authService.getToken();
    return new HttpHeaders({
      Authorization: `Bearer ${token}`,
    });
  }

  /**
   * B-16. Listar, Filtrar y Paginar Asignaciones
   */
  listarPaginado(
    filtro: string,
    estado: string,
    codn: number, // 0 si es "Todos los niveles"
    page: number,
    size: number
  ): Observable<PageResponse<Dicta>> {

    const pageZeroBased = page - 1;
    let params = new HttpParams()
      .set('filtro', filtro)
      .set('estado', estado)
      .set('page', pageZeroBased.toString())
      .set('size', size.toString());

    if (codn > 0) {
      params = params.set('codn', codn.toString());
    }

    return this.http.get<PageResponse<Dicta>>(this.apiURL, {
      headers: this.getAuthHeaders(),
      params: params,
    });
  }

  /**
   * B-16.1. Adicionar Nueva Asignación
   */
  crear(codmat: string, codpar: number, codp: number, gestion: number, login: string): Observable<Dicta> {
    let params = new HttpParams()
      .set('codmat', codmat)
      .set('codpar', codpar.toString())
      .set('codp', codp.toString())
      .set('gestion', gestion.toString())
      .set('login', login);

    return this.http.post<Dicta>(this.apiURL, null, { // Cuerpo vacío
      headers: this.getAuthHeaders(),
      params: params
    });
  }

  /**
   * B-16.3. Eliminar (Baja lógica)
   */
  eliminar(id: Dicta['id']): Observable<void> {
    let params = new HttpParams()
      .set('codmat', id.codmat)
      .set('codpar', id.codpar.toString())
      .set('codp', id.codp.toString())
      .set('gestion', id.gestion.toString());

    return this.http.delete<void>(this.apiURL, {
      headers: this.getAuthHeaders(),
      params: params
    });
  }

  /**
   * B-16. Habilitar (no está en la UI, pero lo implementamos)
   */
  habilitar(id: Dicta['id']): Observable<void> {
    let params = new HttpParams()
      .set('codmat', id.codmat)
      .set('codpar', id.codpar.toString())
      .set('codp', id.codp.toString())
      .set('gestion', id.gestion.toString());

    return this.http.put<void>(`${this.apiURL}/habilitar`, null, { // Cuerpo vacío
      headers: this.getAuthHeaders(),
      params: params
    });
  }
 modificar(
    // Clave Vieja (4 argumentos)
    oldCodmat: string,
    oldCodpar: number,
    oldCodp: number,
    oldGestion: number,
    // Clave Nueva (3 argumentos)
    newCodmat: string,
    newCodpar: number,
    newCodp: number,
    // Login (1 argumento)
    login: string
  ): Observable<Dicta> {

    let params = new HttpParams()
      // Clave Vieja
      .set('oldCodmat', oldCodmat)
      .set('oldCodpar', oldCodpar.toString())
      .set('oldCodp', oldCodp.toString())
      .set('oldGestion', oldGestion.toString())
      // Clave Nueva
      .set('newCodmat', newCodmat)
      .set('newCodpar', newCodpar.toString())
      .set('newCodp', newCodp.toString())
      .set('login', login);

    return this.http.put<Dicta>(this.apiURL, null, { // Llama al endpoint @PutMapping
      headers: this.getAuthHeaders(),
      params: params
    });
  }
 }

File: src/main/java/com/Proyecto/backEnd/service/DictaService.java
package com.Proyecto.backEnd.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import com.Proyecto.backEnd.model.*;
import com.Proyecto.backEnd.repository.*;
import jakarta.persistence.criteria.*;
import jakarta.transaction.Transactional;

@Service
public class DictaService {

    @Autowired
    private DictaRepo dictaRepo;
    @Autowired
    private UsuariosRepo usuariosRepo;
    @Autowired
    private MateriasRepo materiasRepo;
    @Autowired
    private ParalelosRepo paralelosRepo;
    @Autowired
    private PersonalRepo personalRepo;

    /**
     * B-16. Listar, filtrar y paginar
     */
    public Page<DictaModel> listarPaginado(String filtro, String estado, Integer codn, Pageable pageable) {
        
        Specification<DictaModel> spec = (root, query, cb) -> {
            // Hacemos fetch para cargar las entidades relacionadas y evitar N+1 queries
            root.fetch("materia").fetch("nivel");
            root.fetch("profesor");
            root.fetch("paralelo");
            query.distinct(true);
            
            Predicate p = cb.conjunction();

            // Filtro por Estado (Activos/Bajas/Todos)
            if (estado.equals("ACTIVOS")) {
                p = cb.and(p, cb.equal(root.get("estado"), 1));
            } else if (estado.equals("BAJAS")) {
                p = cb.and(p, cb.equal(root.get("estado"), 0));
            }

            // Filtro por Nivel (codn)
            if (codn != null && codn > 0) {
                Join<DictaModel, MateriasModel> materiaJoin = root.join("materia");
                Join<MateriasModel, NivelesModel> nivelJoin = materiaJoin.join("nivel");
                p = cb.and(p, cb.equal(nivelJoin.get("codn"), codn));
            }

            // Filtro por Texto (Nombre/Apellido de Profesor O Nombre de Materia)
            if (filtro != null && !filtro.isEmpty()) {
                String filtroLower = "%" + filtro.toLowerCase() + "%";
                
                Join<DictaModel, PersonalModel> profesorJoin = root.join("profesor");
                Join<DictaModel, MateriasModel> materiaJoin = root.join("materia", JoinType.LEFT); // Usamos LEFT por si el join ya existía

                Predicate filtroProfesor = cb.or(
                    cb.like(cb.lower(profesorJoin.get("nombre")), filtroLower),
                    cb.like(cb.lower(profesorJoin.get("ap")), filtroLower)
                );
                Predicate filtroMateria = cb.like(cb.lower(materiaJoin.get("nombre")), filtroLower);

                p = cb.and(p, cb.or(filtroProfesor, filtroMateria));
            }
            
            return p;
        };
        
        return dictaRepo.findAll(spec, pageable);
    }

    /**
     * B-16.1. Adicionar Nueva Asignación
     */
    public DictaModel crearAsignacion(String codmat, int codpar, int codp, int gestion, String login) {
        
        DictaId id = new DictaId(codpar, codp, codmat, gestion);
        if (dictaRepo.existsById(id)) {
            throw new RuntimeException("Esta asignación (Profesor-Materia-Paralelo) ya existe para esta gestión.");
        }

        // Buscamos las entidades para validarlas
        MateriasModel mat = materiasRepo.findById(codmat).orElseThrow(() -> new RuntimeException("Materia no encontrada"));
        ParalelosModel par = paralelosRepo.findById(codpar).orElseThrow(() -> new RuntimeException("Paralelo no encontrado"));
        PersonalModel prof = personalRepo.findById(codp).orElseThrow(() -> new RuntimeException("Profesor no encontrado"));
        UsuariosModel usu = usuariosRepo.findById(login).orElseThrow(() -> new RuntimeException("Usuario (login) no encontrado"));

        DictaModel nueva = new DictaModel();
        nueva.setId(id);
        nueva.setMateria(mat);
        nueva.setParalelo(par);
        nueva.setProfesor(prof);
        nueva.setUsuario(usu);
        nueva.setEstado(1); // Activo por defecto

        return dictaRepo.save(nueva);
    }
    
    /**
     * B-16.3. Eliminar Asignación (Baja lógica)
     */
    public void eliminarLogico(DictaId id) {
        DictaModel d = dictaRepo.findById(id).orElseThrow(() -> new RuntimeException("Asignación no encontrada"));
        d.setEstado(0); // Damos de baja
        dictaRepo.save(d);
    }

    /**
     * Habilitar Asignación (similar a B-16.3)
     */
    public void habilitar(DictaId id) {
        DictaModel d = dictaRepo.findById(id).orElseThrow(() -> new RuntimeException("Asignación no encontrada"));
        d.setEstado(1); // Reactivamos
        dictaRepo.save(d);
    }
    
    @Transactional
    public DictaModel modificarAsignacion(
            // --- Clave Vieja (para borrar)
            String oldCodmat, int oldCodpar, int oldCodp, int oldGestion,
            // --- Clave Nueva (para crear)
            String newCodmat, int newCodpar, int newCodp, String login) {
        
    	 // 1. Buscar la asignación original usando únicamente la combinación
        //    materia-paralelo-gestión. Así evitamos depender de que el profesor
        //    "viejo" llegue correcto desde la UI.
        DictaModel original = dictaRepo
            .findById_CodmatAndId_CodparAndId_Gestion(oldCodmat, oldCodpar, oldGestion)
            .orElseThrow(() -> new RuntimeException("No se encontró la asignación original para modificar."));

        DictaId idNuevo = new DictaId(newCodpar, newCodp, newCodmat, oldGestion);
        if (dictaRepo.existsById(idNuevo) && !idNuevo.equals(original.getId())) {
            throw new RuntimeException("La asignación nueva ya existe para la gestión seleccionada.");
        }

        // 2. Preparar las entidades nuevas
        MateriasModel mat = materiasRepo.findById(newCodmat).orElseThrow(() -> new RuntimeException("Materia no encontrada"));
        ParalelosModel par = paralelosRepo.findById(newCodpar).orElseThrow(() -> new RuntimeException("Paralelo no encontrado"));
        PersonalModel prof = personalRepo.findById(newCodp).orElseThrow(() -> new RuntimeException("Profesor no encontrado"));
        UsuariosModel usu = usuariosRepo.findById(login).orElseThrow(() -> new RuntimeException("Usuario (login) no encontrado"));

        // 3. Si no cambió la clave primaria, solo actualizamos los vínculos
        if (idNuevo.equals(original.getId())) {
            original.setMateria(mat);
            original.setParalelo(par);
            original.setProfesor(prof);
            original.setUsuario(usu);
            return dictaRepo.save(original);
        }
        // 4. Crear la asignación nueva y recién al final eliminar la original
        DictaModel nueva = new DictaModel();
        nueva.setId(idNuevo);
        nueva.setMateria(mat);
        nueva.setParalelo(par);
        nueva.setProfesor(prof);
        nueva.setUsuario(usu);
        nueva.setEstado(original.getEstado());

        DictaModel guardada = dictaRepo.save(nueva);
        dictaRepo.delete(original);

        return guardada;
    }
}